File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
  public void reorderList(ListNode head) 
    {
        if (head == null || head.next == null) return;

        ListNode slow = head;
        ListNode fast = head;
        
        while(fast!=null && fast.next!=null)
        {
            fast = fast.next.next;
            slow = slow.next;
        }
        // slow 6 8 10

        //
        ListNode secondHalf = reverse(slow.next);//8 6
        slow.next = null;// 2 4 null

        ListNode firstHalf = head;// 2 4 6 8

        while(secondHalf!=null)
        {
            ListNode temp1 = firstHalf.next; //4 6 8
            ListNode temp2 = secondHalf.next; //6
            
            firstHalf.next = secondHalf;//2 8 6
            secondHalf.next = temp1;//8->temp1

            firstHalf = temp1;
            secondHalf = temp2;
        }

    }

    private ListNode reverse(ListNode node)
    {
       ListNode prev = null;
       ListNode next = null;
       ListNode curr = node;

       while(curr!=null)// 1 2 3 4
       {
        next = curr.next;//nx=2 3 4
        curr.next = prev;//1 null
        prev = curr;
        curr = next;
       } 
       return prev;
    }
}

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if(head is None):
            return head
        return self.reverseListHelper(head)
    

    def reverseListHelper(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev: ListNode = None
        curr: ListNode = head
        node: ListNode = None
        # prev = None
        # curr = head

        while curr is not None:
            node = curr.next
            curr.next = prev
            prev = curr
            curr = node
        return prev
        


        